### ISSUE TITLE
Wait for stream to finish sending before shutdown

### ISSUE THREAD

**Issue by matzf** [lucas-clemente/quic-go#3291]
An application using quic-go currently has no reliable way to wait for a/all quic sessions/streams to finish sending before shutting down.
When an application writes on a stream, the data is buffered internally in quic-go, similar TCP. If the application then shuts down (or closes the quic session or closes the UDP socket) before all the buffered data has actually been transmitted, this will simply be lost. 

Side note: with TCP, this issue (usually) does not occur because the kernel keeps working on this send buffer after the application has shut down (and there is also the option `SO_LINGER` to control explicitly blocking `close` calls until all data has been transmitted). In QUIC, this is all in user space, so obviously once the application has shut down, the buffered data will not be transmitted.

Side note: for HTTP/3 this issue does not appear to occur by virtue of how QUIC is used. The server is long-lived to begin with, so any data it sends will reliably be transmitted (but CloseGracefully might be affected). The server usually replies after reading the full request and the (short lived) client then reads the full server reply. Thus, the client's data will be fully transmitted before shutting down.

#### Demonstration / Reproduction:

<details>
<summary>main.go</summary>

```go
package main

import (
	"context"
	"crypto/tls"
	"flag"
	"fmt"
	"io"
	"strings"
	"time"

	"github.com/lucas-clemente/quic-go"
)

func main() {
	serv := flag.Bool("s", false, "Run as server, default run as client")
	certFile := flag.String("cert", "", "Certificat file for server")
	keyFile := flag.String("key", "", "Key file for server")
	flag.Parse()
	var err error
	if *serv {
		err = server(*certFile, *keyFile)
	} else {
		err = client()
	}
	if err != nil {
		fmt.Println("ERROR", err)
	}
}

func server(certFile, keyFile string) error {
	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
	if err != nil {
		return err
	}
	listener, err := quic.ListenAddr(
		"localhost:3333",
		&tls.Config{
			NextProtos:   []string{"test"},
			Certificates: []tls.Certificate{cert},
		},
		nil,
	)
	if err != nil {
		return err
	}
	ctx := context.Background()
	sess, err := listener.Accept(ctx)
	if err != nil {
		return err
	}
	stream, err := sess.AcceptUniStream(ctx)
	if err != nil {
		return err
	}
	stream.SetReadDeadline(time.Now().Add(1 * time.Second)) // make this time out a bit quicker
	n, err := io.Copy(io.Discard, stream)
	fmt.Println("read", n, "from", sess.RemoteAddr(), "err:", err)
	return nil
}

func client() error {
	sess, err := quic.DialAddr(
		"localhost:3333",
		&tls.Config{NextProtos: []string{"test"}, InsecureSkipVerify: true},
		nil,
	)
	if err != nil {
		return err
	}
	stream, err := sess.OpenUniStream()
	if err != nil {
		return err
	}
	data := []byte(strings.Repeat("test_data_", 1000))
	n, err := stream.Write(data)
	if err != nil {
		return err
	}
	fmt.Println("sent", n, "to", sess.RemoteAddr())
	return stream.Close()
}
```
</details>

Run: 
- Start server: `go run main.go -s quic-go/internal/testdata/cert.pem -key quic-go/internal/testdata/priv.key` 
- Start client: `go run main.go`
- Observe that (usually):
  - client says: "sent 10000 to 127.0.0.1:3333`
  - server says eg: "read 2346 from 127.0.0.1:49593 err: deadline exceeded"
    Note that this is issue inherently timing dependent. Sometimes the server reports success (full data read and no error), or reading the full data but a timeout (the FIN for the stream was not sent/received).

Notes:
- If we add a sufficiently long `time.Sleep()` in there after closing the stream in the client, this example will work reliably.
- The example uses a unidirectional stream to make it obvious which way the data flows here. But bidirectional streams are affected just the same.
- In the example, the client is short lived and it's data send stream is cut off. However, a short lived server (i.e. a server sending a response and terminating immediately after closing the stream) would have the same problem. This seems like a less common scenario, but it could still happen (something like `netcat -l`). 

#### Proposal:
Add an API to allow an application to block until all data on a stream / session has been sent before shutting down.
This could be either on the level of a `SendStream` or a `Session` (or both). For the shutdown issue described above, something on the level of the Session might be more convenient. On the other hand, having such a functionality on the level of Streams could facilitate other advanced usage.

Perhaps the most straight forward way to add this to the interface would be:
-  `Session.CloseWithErrorSync(context.Context, ApplicationErrorCode, string) error`: this blocks until all buffered data has been transmitted and then closes the session.
- `Stream.CloseSync(context.Context) error`: close the stream and then wait for all buffered data to be transmitted.

**Comment by marten-seemann** [comment]
I disagree. The way that streams are used is highly dependent on the application, and what works well for one application doesn’t work for the other. 
Introducing an API that breaks in the case of packet loss seems incredibly brittle.

**Comment by guodoliu** [comment]
@marten-seemann Hi, any update on this enhancement? i run into the same issue

**Comment by marten-seemann** [comment]
@nisdas I moved this to a separate issue (#4404), since it's not _quite_ related to the discussion here. Will respond on that issue.

**Comment by sukunrt** [comment]
I like the idea of a notification that the stream has finished sending, but that will not help with connection close. This is because QUIC doesn't specify any behavior with regards to received data on streams on receiving `CONNECTION_CLOSE`. This would mean that even if the quic layer has ACKed the data, remote quic stack might discard the data on receiving a `CONNECTION_CLOSE`.

**Comment by qAison** [comment]
I have the same problem? Is there any good solution now?
Is there a method similar to Stream.Flush()?
